### DISCUSSION TITLE
New feature: Sort columns, search/filter, pagination script

### DISCUSSION THREAD

**Discussion by Zen4All** [zencart/zencart#5781] [Ideas]
These are features that were already present in ZC 2.0. Today I was doing some googling,and came across this cool script that does a lot of stuf we were talking about here on Github.
for the scripts site see https://datatables.net/
The script paginates, sorts tables, and enables dynamically searching on a page. 
I spend about 10 minutes on installation/customization, and already had a pretty workable page.
Maybe it is something we can use, in a next version, or else I'll try to make a mod out of it.
![Sort columns, search/filter, pagination script](https://user-images.githubusercontent.com/2766593/81866892-2a7e0880-9570-11ea-91d1-110a92daf73e.png)

**Comment by torvista** [comment]
When I looked, a while ago, I got the impression that this was the grand-daddy of them all: does all you could want and more....which made me think that it would be overkill for simple sorting....but in reality arguing over a few kb is pointless these days, so why not, throw in the biggest and best!

**Comment by lankeeyankee** [comment]
I love this and definitely would use it. On large sites like mine, this is invaluable and saves so much time and effort. If they don't want it in the core a module would be most welcome.

**Comment by drbyte** [comment]
The datatables script is quite mature and very flexible.
I've used it on some projects unrelated to Zen Cart. It was easy to implement. I later removed it in favor of other approaches that were much lighter-weight and less complex for maintaining.

(Aside: A nice addon is the in-place-editing upgrade. It requires purchasing a license for each site you use it on. As long as an "edit" button is offered in the base implementation, the paid option could remain totally optional.)

One implementation challenge is that it relies strictly on ajax for populating the tables, so back-end work needs to be done both to expose and to secure the related endpoints for the ajax calls. Removing all the query logic from the individual admin pages would make this easier, but really begs for an OOP implementation. We'd been doing a lot of that work in v2.

**Comment by drbyte** [comment]
This may be lighter weight: https://gridjs.io

**Comment by torvista** [comment]
Prompted by noting that the column sorting in customers is not 100% (both links stay as links instead of the selected on being not a link) and it's very inelegant code...anything more suggestions to try out before I have a go at gridjs?

**Comment by torvista** [comment]
I did a method to add jquery tablesorter to admin tables relatively easily (in my Github), but it can only sort the displayed table/on that page, so only valid where the table holds all the results, not spread across pages. I guess that can only be done internally/not by an add-on.

**Comment by zcwilt** [comment]
The plugin manager uses a set of classes to build a table based on a laravel model/builder query
It supports notifiers to adjust the initial query and columns displayed as well allowing for adding action buttons e.g. New/Create
and uses a paginator.

I've recently been updating the code to add filters (e.g show where status =? etc) and I have been working on orderBys over the last couple of days. The current code does not use any javascript, so page rebuilds on filters/order changes. The idea would be to allow queries to be ajax driven sometime in the future.

The main downside (or upside if you are a Laravel fan) is that we would need to create models etc for pages that are being converted.

Hoping to have a demo of this shortly, and maybe convert another page in admin as well.

**Comment by gothstone** [comment]
Was any further progress made by anyone? About to dink around with several report plugins and column sorting options would be nice.
